### ISSUE TITLE
false negative of `semi` rule with `never` option

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#9521]
**Tell us about your environment**

* **ESLint Version:** 4.9.0
* **Node Version:** 8.6.0
* **npm Version:** 5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?**
**Please show your full configuration:**
**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

- [Open the online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyplc2xpbnQgc2VtaTogW2Vycm9yLCBuZXZlcl0sIG5vLWV4dHJhLXNlbWk6IGVycm9yICovXHJcblxyXG5pbXBvcnQgYSBmcm9tIFwiYVwiOyAvL+KGkCBgc2VtaWAgcnVsZSBzaG91bGQgd2FybiB0aGlzIHNlbWljb2xvbi5cclxuWzEsMiwzXS5mb3JFYWNoKGRvU29tZXRoaW5nKVxyXG5cclxuZXhwb3J0IHthfTsgLy/ihpAgYHNlbWlgIHJ1bGUgc2hvdWxkIHdhcm4gdGhpcyBzZW1pY29sb24uXHJcblsxLDIsM10uZm9yRWFjaChkb1NvbWV0aGluZylcclxuXHJcbmZ1bmN0aW9uIHdyYXAoKSB7XHJcblx0cmV0dXJuOyAvL+KGkCBgc2VtaWAgcnVsZSBzaG91bGQgd2FybiB0aGlzIHNlbWljb2xvbi5cclxuXHRbMSwyLDNdLmZvckVhY2goZG9Tb21ldGhpbmcpXHJcbn1cclxuXHJcbndoaWxlICh0cnVlKSB7XHJcbiAgICBicmVhazsgLy/ihpAgYHNlbWlgIHJ1bGUgc2hvdWxkIHdhcm4gdGhpcyBzZW1pY29sb24uXHJcbiAgICBbMSwyLDNdLmZvckVhY2goZG9Tb21ldGhpbmcpXHJcbiAgICBcclxuICAgIGNvbnRpbnVlOyAvL+KGkCBgc2VtaWAgcnVsZSBzaG91bGQgd2FybiB0aGlzIHNlbWljb2xvbi5cclxuICAgIFsxLDIsM10uZm9yRWFjaChkb1NvbWV0aGluZylcclxufVxyXG5cclxuZG87IHdoaWxlKGEpOyAvL+KGkCBgc2VtaWAgcnVsZSBzaG91bGQgd2FybiB0aGlzIHNlbWljb2xvbi5cclxuWzEsMiwzXS5mb3JFYWNoKGRvU29tZXRoaW5nKVxyXG5cclxuY29uc3QgZiA9ICgpID0+IHt9OyAvL+KGkCBgc2VtaWAgcnVsZSBzaG91bGQgd2FybiB0aGlzIHNlbWljb2xvbi5cclxuWzEsMiwzXS5mb3JFYWNoKGRvU29tZXRoaW5nKVxyXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjgsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=).

**What did you expect to happen?**

I expected `semi` rule to warn those semicolons because I'm using `"never"` option.

If a semicolon exists the end of the following statements, I can remove semicolons safely even if the next line starts with `(` or `[`:

- `DoWhileStatement`
- `ReturnStatement[argument=null]`
- `BreakStatement`
- `ContinueStatement`
- `DebuggerStatement`
- `ImportDeclaration`
- `ExportNamedDeclaration[declaration=null]`
- `ExportAllDeclaration`
- Statements that the previous token of the semicolon is the right brace of an arrow function body.

**What actually happened? Please include the actual, raw output from ESLint.**

The `semi` rule does not warn those semicolons.

-----

EDIT: After discussion, I changed this bug report to an enhancement proposal: https://github.com/eslint/eslint/issues/9521#issuecomment-339915724 and https://github.com/eslint/eslint/issues/9521#issuecomment-341389666

**Comment by not-an-aardvark** [comment]
I think some users might prefer to *always* have a semicolon before `(` and `[` for consistency, even if there is no ASI hazard in those cases (e.g. after `do-while`.

cc @feross

**Comment by mysticatea** [comment]
I see. In that case, I'd like to add an option to control (require/disallow) the semicolons before `(`/`[`. Currently, `semi` rule merely ignores such semicolons.

For example:

```js
/*eslint semi: [error, never, { beforeUnreliableLineButSafe: "any" }] */
import f from "f" //← OK either way. This is the current behavior. This is default.
[1,2,3].forEach(f)
```

```js
/*eslint semi: [error, never, { beforeUnreliableLineButSafe: "always" }] */
import f from "f" //← Error: require semi. (note: `semi-style` rule will move the semicolon to the beginning of the next line in auto-fixing.)
[1,2,3].forEach(f)
```

```js
/*eslint semi: [error, never, { beforeUnreliableLineButSafe: "never" }] */
import f from "f"; //← Error: disallow semi.
[1,2,3].forEach(f)
```

But I'm not sure better option name.

----

A side note, `indent` rule seems wrong after `import` declarations.

```js
/*eslint indent: error */
import f from "f"
;[1,2,3].forEach(f) //← Error: expected indentation of 4 spaces but found 0. (indent)
```

**Comment by mysticatea** [comment]
I'd like to champion this new option.

I think the option is helpful for consistency of semicolons.

- `{ beforeUnreliableLineButSafe: "always" }` requires semicolons before `[` or `(` always.
- `{ beforeUnreliableLineButSafe: "never" }` disallow extra semicolons.
- `{ beforeUnreliableLineButSafe: "any" }` would keep the current bahavior. This is default.

I'll implement this if this is accepted.

@eslint/eslint-team What do you think?

**Comment by Aladdin-ADD** [comment]
I like the option, but the ASI is very subtle, we might have to consider some more cases:
* before these characters:`(`, `[`, `+`, `-`, `/`.
```js
// ++
a
++
b

// /
a = b
/xxx/i.test(str)
```
* for-loop:
```js
for (var i = 0; i < len; i++); // this semi cannot be omitted
```
any others?

**Comment by mysticatea** [comment]
@Aladdin-ADD 

> always have a semi before ( and [ is more straightforward

Yeah, the `{ beforeUnreliableLineButSafe: "always" }` will support it. The current `semi` rule allows missing semicolons before `(` and `[`.

----

https://github.com/eslint/eslint/commit/08b654c85cc57669c54ab46153ef3fee39163be3 is a reference implementation.
